[RISC-V] Remove unnecessary assertion from emitOutputInstr

tomeksowi opened the pull request:
Remove unnecessary assertion introduced in #98226. It fired a false positive when W^X was enabled.

Part of #84834
cc @wscho77 @HJLeee @clamp03 @JongHeonChoi @t-mustafin @gbalykov @viewizard @ashaurtaev @brucehoult @sirntar @yurai007 @Bajtazar @bartlomiejko

tomeksowi committed:
Remove unnecessary assertion

It fired a false positive when W^X was enabled

tomeksowi commented on src/coreclr/jit/emitriscv64.cpp:
```diff
@@ -2110,19 +2110,6 @@ void emitter::emitJumpDistBind()
- *  Emit a 32-bit RISCV64 instruction
- */
-
-unsigned emitter::emitOutput_Instr(BYTE* dst, code_t code) const
-{
-    assert(dst != nullptr);
-    assert(sizeof(code_t) == 4);
-    memcpy(dst + writeableOffset, &code, sizeof(code_t));
-    return sizeof(code_t);
-}
```
One reason was name confusion, we had emitOutputInstr and emitOutput_Instr which did different things under almost the same name. Another is to reuse common code rather than maintaining our own helpers that do the same. Do you see any problem with emitOutputLong for RISC-V?

clamp03 replied:
Thank you for the comment.
- I think `emitOutput_Instr` is more intuitive than `emitOutputLong`. Actually, I thought `emitOutputLong` writes long value (8 bytes) into dst.
- ARM64 and Loongarch64 use `emitOutput_Instr`. XARCH use `emitOutputLong` and others such as `emitOutputByte`, `emitOutputWord` and so on. I want RISC-V looks similar to ARM64 and LoongArch64 than the others because RISC-V referenced many things from ARM64 and LoongArch64.

tomeksowi replied:
You're right, only xarch uses it. OK let's keep emitOutput_Instr. Thank you.

clamp03 replied:
How do you think about changing implementation from 'memcpy' to storing value to the pointer like arm64?
I think the function will be called very many and 'memcpy' overhead is bigger than storing. I think such overhead is unnecessary.

tomeksowi replied:
No problem

clamp03 replied:
Thank you

yurai007 replied:
As side note/nit: when N is very small and known at compile time usually memcpy(N)/memset(N) are lowered by 
Clang/GCC (-O2/-O3) to equivalent loads/stores so it's likely that both forms are fine (at least on Linux :).

tomeksowi replied:
I know but:
* it's lowered to a sequence of sb + srli vs. a single sw
* no need to pay the tax in debug mode

tomeksowi committed:
Use uint32 write instead of memcpy to be faster

There shouldn't be a problem when we introduce compressed instructions, VF2 supports unaligned stores (sw)
Also, replace compile-time conditions with static_asserts.

bartlomiejko reviewed (approved):
LGTM

jakobbotsch commented on src/coreclr/jit/emitriscv64.cpp:
```diff
@@ -2118,8 +2118,8 @@ void emitter::emitJumpDistBind()
-    assert(sizeof(code_t) == 4);
-    memcpy(dst + writeableOffset, &code, sizeof(code_t));
+    static_assert(sizeof(code_t) == 4, "code_t must be 4 bytes");
+    *(code_t*)(dst + writeableOffset) = code;
```
The buffer is not guaranteed to be aligned (for example in prejit it is just the result of a normal byte array allocation). Are unaligned writes on RV64 allowed?

tomeksowi replied:
VisionFive 2 allows it and I think most 64-bit implementations do. The [RVA20U64 profile mandates it](https://github.com/riscv/riscv-profiles/blob/54717ad320c37f27d235483f23e3f0a805235c37/profiles.adoc?plain=1#L529-L538) but currently we compile for RV64GC ISA which does allow trapping and I don't see `-mno-strict-align` anywhere so let's play it safe. Thank you for noticing.

This could be \#ifdef'ed in the future when we get the [recommended macros in clang](https://github.com/llvm/llvm-project/pull/65756), I don't see them in 17.0.6 yet.

brucehoult replied:
> RV64GC ISA which does allow trapping

RV64GC and RVA20U64 are the same thing. RVA20 is just a new, extensible, name for RV64GC. RISC-V has always required User-mode misaligned access to work. It *may* trap, but it must be trap-and-emulate and work in the end (but slowly).  When RVA20U64 says "Even though mandated, misaligned loads and stores might execute extremely slowly." what they mean is it may be trap-and-emulate. EXTREMELY slowly. There is no change in this from RV64GC.

> The buffer is not guaranteed to be aligned (for example in prejit it is just the result of a normal byte array allocation).

It would be very unusual in any memory allocator for any object to not be aligned to at least 4 bytes, if not 16. Is this really likely?

jakobbotsch replied:
During prejit (where code is being emitted into an object file in the end) this is just the result of a normal byte array allocation coming from the EE side. The buffer is likely to be aligned, and in practice it will be, but it is not a guarantee provided by the JIT-EE interface. If you want to allocate a temporary buffer yourself and guarantee its alignment, then you can do that.

> It seems to me this is pretty performance-critical code, and really doesn't want to suffer calling memcpy() -- which the compiler can't optimise out as it knows casting a BYTE* to a code_t* does not make an aligned pointer.

Writing the code to the buffer is a drop in the bucket compared to everything else that goes on during codegen. Each IR node that resulted in the instructions written are orders of magnitudes larger than the corresponding final bytes written to the buffer.

The compiler is perfectly capable of optimizing out memcpy calls with constant size. That is optimized by all compilers I know of. @tomeksowi pointed out above that it is turned into some sequence of sb + srl on RISCV. It does not turn into a call.

tomeksowi replied:
> @tomeksowi pointed out above that it is turned into some sequence of sb + srli on RISCV

[Yes](https://godbolt.org/z/P55aeveMv), and given we will support compressed instructions eventually, in the long run the improvement will be shaving off 4 instructions (2x sh + 1x srli vs. 4x sb + 3x srli now), that's if we guarantee 2-byte alignment ourselves. IMO juice not worth the squeeze.

tomeksowi committed:
Bring back memcpy because plain RV64 ISA allows trapping on misaligned load/stores
